Userland/nc.cpp

```diff
@@ -11,68 +13,174 @@
 
 #define max(a, b) ((a > b) ? a : b)
 
+void exit_with_usage(int rc)
+{
+    fprintf(stderr, "Usage: nc [-l] [-v] [-h] [-N] [-s source] [-p source_port] [destination] [port]\n");
+    exit(rc);
+}
+
 int main(int argc, char** argv)
 {
-    if (argc < 3) {
-        fprintf(stderr, "Usage: nc <host> <port>\n");
-        return -1;
+    bool should_listen = false;
+    bool verbose = false;
+    char *source_addr = nullptr;
+    int source_port = 0;
+    bool should_close = false;
+
+    int opt;
+    while ((opt = getopt(argc, argv, "hlvs:p:N")) != -1) {
+        switch (opt) {
+        case 'h':
+            exit_with_usage(0);
+            break;
+        case 'l':
+            should_listen = true;
+            break;
+        case 'v':
+            verbose = true;
+            break;
+        case 's':
+            source_addr = strdup(optarg);
+            break;
+        case 'p':
+            source_port = atoi(optarg);
+            break;
+        case 'N':
+            should_close = true;
+            break;
+        }
     }
 
-    const char* addr_str = argv[1];
-    int port = atoi(argv[2]);
+    int fd;
 
-    int fd = socket(AF_INET, SOCK_STREAM, 0);
-    if (fd < 0) {
-        perror("socket");
-        return 1;
-    }
+    if (should_listen) {
+        int listen_fd = socket(AF_INET, SOCK_STREAM, 0);
+        if (listen_fd < 0) {
+            perror("socket");
+            return 1;
+        }
 
-    struct timeval timeout {
-        3, 0
-    };
-    int rc = setsockopt(fd, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));
-    if (rc < 0) {
-        perror("setsockopt");
-        return 1;
-    }
-    rc = setsockopt(fd, SOL_SOCKET, SO_SNDTIMEO, &timeout, sizeof(timeout));
-    if (rc < 0) {
-        perror("setsockopt");
-        return 1;
-    }
+        struct sockaddr_in sa;
+        memset(&sa, 0, sizeof sa);
+        sa.sin_family = AF_INET;
+        sa.sin_port = htons(source_port);
+        sa.sin_addr.s_addr = htonl(INADDR_ANY);
+        if (source_addr) {
+            if (inet_pton(AF_INET, source_addr, &sa.sin_addr) < 0) {
+                perror("inet_pton");
+                return 1;
+            }
+        }
 
-    struct sockaddr_in dst_addr;
-    memset(&dst_addr, 0, sizeof(dst_addr));
+        if (bind(listen_fd, (struct sockaddr*)&sa, sizeof(sa)) == -1) {
+            perror("bind");
+            return 1;
+        }
 
-    dst_addr.sin_family = AF_INET;
-    dst_addr.sin_port = htons(port);
-    rc = inet_pton(AF_INET, addr_str, &dst_addr.sin_addr);
-    if (rc <= 0) {
-        perror("inet_pton");
-        return 1;
-    }
+        if (listen(listen_fd, 1) == -1) {
+            perror("listen");
+            return 1;
+        }
+
+        char addr_str[100];
```

Comment:
Maybe we should implement `inet_ntoa()` to make this kinda gunk slightly less necessary.
Or maybe it easier to use e.g `AK::IPv4Address` here.
Anyways, this is all good :)